Configuration to require fixed_in_version

paspiz85 opened the pull request:
I would need a mechanism to ensure that 'fixed_in_version' are always filled in before the bug reaches the Resolved state. The reason is to populate the changelog for a specific project.

Resolving issues having a target version should set the fixed in version by default.

Fixes [#21734](https://mantisbt.org/bugs/view.php?id=21734)

dregad commented on core/bug_api.php:
```diff
@@ -508,6 +508,16 @@ public function validate( $p_update_extended = true ) {
+				error_parameters( lang_get( 'fixed_in_version' ) );
+				trigger_error( ERROR_EMPTY_FIELD, ERROR );
+			}
+		}
```
There is some duplicated logic here vs changes in bug_change_status_page.php, see if it would be possible avoid that by moving that to an API function somehow.

dregad replied:
> the main check is server-side 

Yes I understand that, but you are not addressing my point about code duplication. 

It's not blocking, but if you are not going to do something about it, then at least a comment should be added here as well as in common.js, to warn future developers that any changes to the logic must be applied to the other location.

dregad commented on config_defaults_inc.php:
```diff
@@ -2299,6 +2299,18 @@
 
+/**
+ * In RESOLVED state fixed_in_version default from target_version.
+ * @global integer $g_bug_fixed_in_version_from_target_version
```
Please use `int` instead of `integer`

dregad reviewed (commented):
What happens if the selected target version is not (no longer) valid for the given project at the time the Issue is resolved ? For example, target version was set to "v1" 2 years ago, "v1" was released without fixing the Issue, and it is finally resolved while "v2" is the current "next" version ?
- set fixed_in_version to v1
- don't set fixed_in_version 
- throw error
- ...

Consider usage via REST or SOAP API, as well as from group action - we need to ensure consistent behavior across the board.

Please break long lines between 80 and max 120 chars.

paspiz85 commented:
@dregad I fixed logic to require fixed_in_version, but haven't a good solution to initialize with target_version then I discarded that logic (it wasn't a required functionality)

dregad reviewed (changes requested):
With this reduced scope, this PR no longer qualifies as a fix for [#21734](https://mantisbt.org/bugs/view.php?id=21734). It also does not match your own [#34407](https://mantisbt.org/bugs/view.php?id=34407). Please reference an appropriate issue in the tracker.

I performed a quick test, but it does not work as expected.

If you edit an issue and set status to resolved, you get _A necessary field "Fixed in Version" was empty_ regardless of selected Resolution. Did you read my earlier review note https://github.com/mantisbt/mantisbt/pull/1998#discussion_r1574463015 
> Additionally, you need to consider that a bug may be in RESOLVED status without having FIXED resolution (e.g. duplicate, no change required, etc) so it's necessary to consider bug_resolution_fixed_threshold

I don't understand the need for the new `fixed_in_version_required_status` hidden input; if it's just to pass value of bug_resolution_fixed_threshold to the javascript, then it's probably better handled with a data attribute. 

Additionally, note that technically more than one Resolution could be considered "fixed" (read PHPDoc for $g_bug_resolution_fixed_threshold) so I'm not sure that referring to a single value is appropriate.

The javascript does not work in edit page.

I did not test this, but again please make sure we have consistent behavior with SOAP/REST API.

dregad commented on core/bug_api.php:
```diff
@@ -508,6 +508,16 @@ public function validate( $p_update_extended = true ) {
+			$t_resolved = config_get( 'bug_resolved_status_threshold', null, null, $this->project_id );
+			$t_closed = config_get( 'bug_closed_status_threshold', null, null, $this->project_id );
+			$t_resolution_fixed = config_get( 'bug_resolution_fixed_threshold', null, null, $this->project_id );
+			if ( $this->status >= $t_resolved && ( $this->status < $t_closed || !$t_bug_resolution_is_fixed ) ) {
```
$t_bug_resolution_is_fixed variable is not defined - did you mean $t_resolution_fixed ?.

paspiz85 commented:
I confused issues and agree: this PR can fix #34407 but only for changelog part.
I think that this PR can be closed because no longer qualifies as a fix for #21734.
Do I need to split issue #34407 in order to continue my work?

Regarding the changes:
- I understood your review note #1998 and now I fixed my code to use (I hope correctly) bug_resolution_fixed_threshold
- Included edit page (in this case I need to also check status changing)
- SOAP/REST API don't use validation bug_api then require an extra check. I will add in function mc_issue_update (mc_issue_api.php)

paspiz85 commented:
Without modifications, via SOAP I receive:
`<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/">
   <SOAP-ENV:Body>
      <SOAP-ENV:Fault>
         <faultcode>SOAP-ENV:Server</faultcode>
         <faultstring>Error Type: APPLICATION ERROR #11,
Error Description: Il campo obbligatorio "Assegnata alla versione" era vuoto.  Ricontrolla le informazioni che hai inserito.</faultstring>
      </SOAP-ENV:Fault>
   </SOAP-ENV:Body>
</SOAP-ENV:Envelope>`
This message is due to mc_error_handler that handle E_USER_ERROR as Server SOAP fault.

dregad commented:
Sorry I've been busy. Haven't had time to review changes submitted since my last review yet, will do as time allows, but just wanted to give you feedback on your questions

> I think that this PR can be closed because no longer qualifies as a fix for #21734.

Not necessary to close this PR - just make sure it references the correct issue(s), i.e. edit https://github.com/mantisbt/mantisbt/pull/1998#issue-2255916985

> Do I need to split issue [#34407](https://mantisbt.org/bugs/view.php?id=34407) in order to continue my work?

Since this is now only covering the Changelog (fixed_in_version), then I would propose to simply edit the summary and description of #34407 (which I would have to do for you) to reflect the new, reduced scope. 

I'm not sure if a separate issue for Roadmap (target_version) is needed, or already covered (more or less) by other, existing issues. I let you decide; if yes, then you can clone [#34407](https://mantisbt.org/bugs/view.php?id=34407) and edit to submit a new Issue.

paspiz85 commented:
I updated issue in PR first comment (now issue is https://mantisbt.org/bugs/view.php?id=34429).
I also checked SOAP/REST services and added a logic to intercept validation and return the correct message

dregad reviewed (changes requested):
Hello @paspiz85 

Apologies for the long delay in responding to you, I only found today some time to test this.

There are still some things that are not quite right.

Firstly, keep in mind that _$g_bug_resolved_status_threshold_ is a **threshold**, meaning that an issue is considered as resolved when its status is equal **or higher**; I already mentioned this in https://github.com/mantisbt/mantisbt/pull/1998#discussion_r1574463015. 

This means that the _fixed_in_version_ field is not only available in Resolved state, but also in later statuses (e.g. CLOSED with default settings). I would therefore expect the same validation rules to apply whenever an issue is considered resolved, i.e. if going straight from ASSIGNED to CLOSED, the version should also be required. Currently with your patch it is not.

Secondly, in the same comment referenced above, I also mentioned
> a bug may be in RESOLVED status without having FIXED resolution (e.g. duplicate, no change required, etc) so it's necessary to consider bug_resolution_fixed_threshold

Currently this does not work as expected. With status = RESOLVED, the mandatory check is active regardless of the selected resolution. Please read the PHPDoc for *$g_bug_resolution_fixed_threshold* and the related *$g_bug_resolution_not_fixed_threshold* configs. This means that fixed_in_version should be required when 
*fixed_threshold <= resolution < not_fixed_threshold*.

To fix this in the most consistent and backwards-compatible possible way, I would recommend to rely on *helper_call_custom_function()* to call `changelog_include_issue` https://github.com/mantisbt/mantisbt/blob/cd1f29156539d21f52df7c5c05aed246348b7d75/core/custom_function_api.php#L71-L77.

![image](https://github.com/user-attachments/assets/681cc60c-6077-428c-948b-8822488752b4)

Also a small visual nitpick, the spacing between the mandatory mark and the label is not right on bug_update_page.php (look at Summary field for reference):
![image](https://github.com/user-attachments/assets/2d60def3-45b8-429d-b89a-263484d9398b)

Note that I only tested the GUI (JavaScript), assuming that the same logic was used for back-end (including API) updates.

dregad commented:
> I can use changelog_include_issue but only server-side (in Javascript I need to check resolution value agains the two threshold manually), I would prefer have same checks without using changelog_include_issue

The problem is, if you do that then anybody who defined an override for that custom function will have inconsistent behavior between the client and server validation, which IMO is not acceptable.

Client-side validation is nice and more user-friendly, but unfortunately MantisBT was not designed for that. Maybe an AJAX ?

paspiz85 commented:
I understand the problems, and with these functions, it’s very important to preserve server behavior.

I fixed the server behavior and removed all client-side behavior to ensure logical consistency.

Adding some client-side logic to create a more user-friendly interface could be useful, but unfortunately, I don't have enough time to set up an AJAX-based solution.

dregad reviewed (changes requested):
Hello @paspiz85 

I just tested this, but unfortunately it does not work as expected, and I am sorry but it's a bit of my fault as I pushed you to use existing custom function logic.

The problem is that *custom_function_default_changelog_include_issue()* receives an *issue ID* and uses it to retrieve information from the database, but we call it from BugData::validate() when the data has not yet been saved. Consequently, the test does not give the expected result, because we check the Issue's stored Resolution (e.g. OPEN == 10) instead of the new one (e.g. FIXED).

My first thought was to move the default comparison logic in custom_function_default_changelog_include_issue() to a helper function receiving a BugData object as parameter, or maybe even better a new BugData method.

The problem with this approach is that we cannot execute user-custom logic in custom_function_override_changelog_include_issue(), and I can't think of a good way to achieve that without breaking backwards-compatibility. 

If we accept the BC-break, we could change the parameter to the custom function from int to BugData. That would of course affect anyone who actually implemented an override, forcing them to adapt their code, but I'm sure most people stick to the default so the impact is probably not so wide.

Any better ideas ?
